File: src/Site/Tankbar.php
<?php

declare(strict_types=1);

namespace MetalLE\Site;

class Tankbar implements Site
{
    public function crawl(): Location
    {
        $events = [];

        $plainHTML = file_get_contents('https://tankbar-leipzig.de/tankevents/');
        if (preg_match_all(
                '#<span class="elementor-icon-list-text">(.*)</span>#siU',
                $plainHTML,
                $matches,
                PREG_SET_ORDER
            ) > 0) {
            foreach ($matches as $match) {
                $txt = $match[1];
                $parts = explode(' : ', $txt);
                $date = $parts[0];
                $date = substr($date, 0, -2) . '20' . substr($date, -2);

                $events[] = new Event(
                    html_entity_decode($parts[1], ENT_HTML5),
                    new \DateTimeImmutable($date),
                );
            }
        }

        return new Location(
            'tb',
            'TankBar Leipzig',
            'https://tankbar-leipzig.de/',
            'img/tankbar.svg',
            $events,
        );
    }

}

File: src/site/tankbar.rs
use std::borrow::Borrow;

use html_escape::decode_html_entities;
use regex::Regex;

use crate::tools::date::parse_short_date;
use crate::{Event, Location, Site, HTTP};

pub(crate) struct Tankbar<'l> {
    location: Location<'l, 'l, 'l>,
}

impl Tankbar<'_> {
    pub fn new() -> Self {
        Self {
            location: Location {
                slug: "tb",
                name: "TankBar Leipzig",
                website: "https://tankbar-leipzig.de/",
            },
        }
    }
}

impl Site for Tankbar<'_> {
    fn get_location(&self) -> &Location {
        self.location.borrow()
    }

    fn fetch_events(&self, http: &HTTP) -> Vec<Event> {
        let mut result = Vec::new();

        let html = http.get("https://tankbar-leipzig.de/tankevents/");
        let reg = Regex::new("(?is)<span class=\"elementor-icon-list-text\">(?P<date>\\d\\d\\.\\d\\d\\.)(?P<year>\\d\\d) : (?P<name>.*?)</span>").unwrap();
        for capture in reg.captures_iter(html.as_str()) {
            let name = capture.name("name").unwrap().as_str();
            if name.contains("Metal") || name.contains("Rock") {
                let name = name.replace("<br>", "");
                let date = capture.name("date").unwrap().as_str();
                let year = capture.name("year").unwrap().as_str();

                result.push(Event::new(
                    decode_html_entities(name.trim()).to_string(),
                    parse_short_date(&*(date.to_owned() + "20" + year)).and_hms(0, 0, 0),
                    self.location.borrow(),
                    "https://tankbar-leipzig.de/tankevents/".to_string(),
                    None,
                ));
            }
        }

        result
    }
}
